include/clang/StaticAnalyzer/Core/BugReporter/InternalInfo.h

```diff
@@ -0,0 +1,99 @@
+//===- InternalInfo.h -------------------------------------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the InternalInfo interfaces.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_STATICANALYZER_CORE_BUGREPORTER_INTERNALINFO_H
+#define LLVM_CLANG_STATICANALYZER_CORE_BUGREPORTER_INTERNALINFO_H
+
+#include "clang/StaticAnalyzer/Core/PathSensitive/ProgramState.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+
+namespace clang {
+namespace ento {
+
+class InternalInfo {
+public:
+  enum Kind { None, Constraint, State};
+
+private:
+  const Kind kind;
+  static constexpr Kind getStaticKind() { return None; }
+
+public:
+  InternalInfo(Kind kind) : kind(kind) {}
+  virtual ~InternalInfo() = default;
+
+  Kind getKind() const { return kind; }
+
+
+  static bool isKind(const InternalInfo& Info) {
+    return Info.getKind() == getStaticKind();
+  }
+
+  template <typename T>
+  T* getAs() {
+    if (!T::isKind(*this))
+      return nullptr;
+    return static_cast<T*>(this);
+  }
+
+  using StringPrinter = llvm::raw_ostream&(llvm::raw_ostream&, StringRef);
+
+  virtual void print(llvm::raw_ostream& Out,
+                     StringPrinter Transform,
+                     StringRef NewlinePrefix = "",
+                     StringRef NewlinePostfix = "\n") = 0;
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+  LLVM_DUMP_METHOD virtual void dump() = 0;
+  LLVM_DUMP_METHOD virtual void dumpToStream(llvm::raw_ostream& Out) = 0;
+#endif
+};
+
+class InternalInfoMap : private std::map<InternalInfo::Kind, std::unique_ptr<InternalInfo>> {
```

Comment:
Please add some comments why you use inheritance and that you want to ensure that each Kind must have only one object.